app/controllers/password_resets_controller.rb

```diff
@@ -0,0 +1,58 @@
+class PasswordResetsController < ApplicationController
+  require_unauthenticated_access
+  before_action :ensure_manual_login_allowed
+
+  layout "public"
+
+  def new
+  end
+
+  def create
+    user_agent = request.env['HTTP_USER_AGENT']
+    os = get_item_in_str(user_agent, KNOWN_OPERATING_SYSTEMS) || "unknown operating system"
+    browser = get_item_in_str(user_agent, KNOWN_BROWSERS) || "unknown browser"
```

Comment:
there's probably a better place for these, but maybe I should just use this?: https://github.com/ua-parser/uap-ruby

Reply:
just added these as methods to the request object